ksql-engine/src/main/java/io/confluent/ksql/metrics/ConsumerCollector.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.metrics;
+
+import io.confluent.common.metrics.KafkaMetric;
+import io.confluent.common.metrics.MetricName;
+import io.confluent.common.metrics.Metrics;
+import io.confluent.common.metrics.Sensor;
+import io.confluent.common.metrics.stats.Rate;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class ConsumerCollector {
+
+  final Map<String, Counter> topicCounters = new HashMap<>();
+  private Metrics metrics;
+  private String id;
+
+  public ConsumerCollector(Metrics metrics) {
+    this.metrics = metrics;
+  }
+
+
+  public ConsumerRecords onConsume(ConsumerRecords records) {
+    if (id == null) {
+      throw new RuntimeException(new IllegalStateException("Cannot run without 'id' being set, use ConsumerCollector.configure(map)"));
+    }
+
+    collect(records);
+    return records;
+  }
+
+  private void collect(ConsumerRecords consumerRecords) {
+    Stream<ConsumerRecord> stream = StreamSupport.stream(consumerRecords.spliterator(), false);
+    stream.forEach((record) -> {
+      topicCounters.computeIfAbsent(record.topic(), k -> new Counter(id, metrics)).increment(record);
+    });
+  }
+  public void configure(Map<String, ?> map) {
+    this.id = (String) map.get(ConsumerConfig.GROUP_ID_CONFIG);
+  }
+
+  public void close() {
+  }
+
+  public String statsForTopic(String topic) {
+    return topicCounters.containsKey(topic) ? "Consumer:" + topicCounters.get(topic).toString() : "";
+  }
+
+  public static class Counter {
+    private final String id;
+    private final Metrics metrics;
+    private Sensor sensor;
+    private KafkaMetric rate;
+
+    public Counter(String id, Metrics metrics) {
+      this.id = id;
+      this.metrics = metrics;
+    }
+
+    Map<Integer, PartitionCounter> partitionCounter = new HashMap<>();
+    long records;
+    long valueBytes;
+
+    public void increment(ConsumerRecord record) {
+      records +=1;
+      valueBytes += record.serializedValueSize();
+
+      if (sensor == null) {
+        String name = "consumer-" + id + "-" +record.topic() + "-" + record.partition();
+        sensor = metrics.sensor(name);
+        System.out.println("Registering:" + name);
+        new RuntimeException("Register:" + name).printStackTrace();
+        MetricName producerRate = new MetricName(name + "-rate-per-sec", name, "consumer-stats");
```

Comment:
looks like you mean `consumerRate` here. just goes to show that there is a strong overalp between this `Counter` and the `Counter` in teh `ProducerCollector`. We should look to drop the repetition.

Reply:
Yeah there def is. Could we also share the `Counter` class with the `ProducerCollector`?

Reply:
I realise this - the problem is the source collection entity is different, ProducerRecord versus ConsumerRecords. I could inject an external iterator to process them and call down to a counter, the other problem is that both of the data types contain different data which doesnt necessarily overlap. i.e. I dont know the size of ProducerRecord.bytes, whereas the consumer record we do - this means it can be used to track bandwidth as opposed to records.